File: backend/Api/TimeLeave/TimeLeave.Entity/Models/TimeLeaveModels.cs
using TimeLeave.Entity.Enums;

namespace TimeLeave.Entity.Models;

public record LeaveTypeDto(
    long LeaveTypeId,
    string Code,
    string Name,
    bool IsPaid,
    bool IsHalfDayAllowed,
    decimal? IsAttachmentRequiredAboveDays,
    Gender? Gender,
    bool IsActive
);

public record CreateLeaveTypeRequest(
    string Code,
    string Name,
    bool IsPaid,
    bool IsHalfDayAllowed,
    decimal? IsAttachmentRequiredAboveDays,
    Gender? Gender
);

public record LeavePolicyDto(
    long LeavePolicyId,
    long LeaveTypeId,
    string? LeaveTypeName,
    long? GradeId,
    long? WorkLocationId,
    DateOnly EffectiveFrom,
    decimal AnnualQuota,
    AccrualKind AccrualKind,
    bool IsProratedOnJoining,
    CarryForwardKind CarryForwardKind,
    decimal? MaxCarryForward,
    decimal? MaxEncashPerYear,
    decimal? MinDaysPerApplication,
    decimal? MaxDaysPerApplication,
    int NoticeDays,
    bool IsSandwichRule,
    bool CanApplyInProbation
);

public record CreateLeavePolicyRequest(
    long LeaveTypeId,
    long? GradeId,
    long? WorkLocationId,
    DateOnly EffectiveFrom,
    decimal AnnualQuota,
    AccrualKind AccrualKind,
    bool IsProratedOnJoining,
    CarryForwardKind CarryForwardKind,
    decimal? MaxCarryForward,
    decimal? MaxEncashPerYear,
    decimal? MinDaysPerApplication,
    decimal? MaxDaysPerApplication,
    int NoticeDays,
    bool IsSandwichRule,
    bool CanApplyInProbation
);

public record LeaveBalanceDto(
    long LeaveBalanceId,
    long EmployeeId,
    long LeaveTypeId,
    string LeaveTypeCode,
    string LeaveTypeName,
    int LeaveYear,
    decimal Opening,
    decimal Accrued,
    decimal Taken,
    decimal Encashed,
    decimal Lapsed,
    decimal Adjusted,
    decimal Available
);

public record ApplyLeaveRequest(
    long EmployeeId,
    long LeaveTypeId,
    DateOnly FromDate,
    DateOnly ToDate,
    LeaveHalf FromHalf,
    LeaveHalf ToHalf,
    string Reason,
    string? AttachmentKey
);

public record LeaveApplicationDto(
    long LeaveApplicationId,
    long EmployeeId,
    long LeaveTypeId,
    string LeaveTypeCode,
    string LeaveTypeName,
    DateOnly FromDate,
    DateOnly ToDate,
    LeaveHalf FromHalf,
    LeaveHalf ToHalf,
    decimal Days,
    string Reason,
    string? AttachmentKey,
    LeaveStatus LeaveStatus,
    ApprovalStatus ApprovalStatus,
    string? CurrentStepLabel,
    long? CurrentApproverEmployeeId
);

public record ApplyEncashmentRequest(
    long EmployeeId,
    long LeaveTypeId,
    int LeaveYear,
    decimal Days
);

public record LeaveEncashmentDto(
    long LeaveEncashmentId,
    long EmployeeId,
    long LeaveTypeId,
    string LeaveTypeCode,
    string LeaveTypeName,
    int LeaveYear,
    decimal Days,
    EncashmentStatus EncashmentStatus,
    ApprovalStatus ApprovalStatus,
    string? CurrentStepLabel
);

public record ShiftDto(
    long ShiftId,
    string Code,
    string Name,
    TimeOnly StartTime,
    TimeOnly EndTime,
    int BreakMinutes,
    int GraceInMinutes,
    int GraceOutMinutes,
    int HalfDayBelowMinutes,
    int AbsentBelowMinutes,
    bool IsNightShift,
    bool IsActive
);

public record CreateShiftRequest(
    string Code,
    string Name,
    TimeOnly StartTime,
    TimeOnly EndTime,
    int BreakMinutes,
    int GraceInMinutes,
    int GraceOutMinutes,
    int HalfDayBelowMinutes,
    int AbsentBelowMinutes,
    bool IsNightShift
);

public record WeeklyOffPolicyDto(
    long WeeklyOffPolicyId,
    string Name,
    WeeklyOffKind MondayRule,
    WeeklyOffKind TuesdayRule,
    WeeklyOffKind WednesdayRule,
    WeeklyOffKind ThursdayRule,
    WeeklyOffKind FridayRule,
    WeeklyOffKind SaturdayRule,
    WeeklyOffKind SundayRule,
    string? AlternateWeeks,
    bool IsActive
);

public record CreateWeeklyOffPolicyRequest(
    string Name,
    WeeklyOffKind MondayRule,
    WeeklyOffKind TuesdayRule,
    WeeklyOffKind WednesdayRule,
    WeeklyOffKind ThursdayRule,
    WeeklyOffKind FridayRule,
    WeeklyOffKind SaturdayRule,
    WeeklyOffKind SundayRule,
    string? AlternateWeeks
);

public record HolidayListDto(
    long HolidayListId,
    string Name,
    long WorkLocationId,
    int CalendarYear,
    int MaxOptionalPerYear,
    List<HolidayDto> Holidays
);

public record CreateHolidayListRequest(
    string Name,
    long WorkLocationId,
    int CalendarYear,
    int MaxOptionalPerYear
);

public record HolidayDto(
    long HolidayId,
    long HolidayListId,
    DateOnly HolidayDate,
    string Name,
    bool IsOptional
);

public record CreateHolidayRequest(
    DateOnly HolidayDate,
    string Name,
    bool IsOptional
);

public record ShiftRosterDto(
    long ShiftRosterId,
    long EmployeeId,
    DateOnly FromDate,
    DateOnly ToDate,
    long ShiftId,
    string? ShiftName,
    long WeeklyOffPolicyId,
    string? WeeklyOffPolicyName
);

public record AssignRosterRequest(
    long EmployeeId,
    DateOnly FromDate,
    DateOnly ToDate,
    long ShiftId,
    long WeeklyOffPolicyId
);

public record PunchDto(
    long PunchId,
    long EmployeeId,
    DateTimeOffset PunchedAt,
    PunchSource PunchSource,
    string? DeviceCode,
    decimal? Latitude,
    decimal? Longitude,
    bool? IsInsideFence
);

public record RecordPunchRequest(
    long EmployeeId,
    DateTimeOffset PunchedAt,
    PunchSource PunchSource,
    string? DeviceCode,
    decimal? Latitude,
    decimal? Longitude
);

public record BiometricPunchImportItem(
    string DeviceUserId,
    DateTimeOffset PunchedAt,
    string? DeviceCode
);

public record BiometricPunchImportRequest(
    string DeviceCode,
    List<BiometricPunchImportItem> Punches
);

public record DailyAttendanceDto(
    long DailyAttendanceId,
    long EmployeeId,
    DateOnly AttendanceDate,
    long? ShiftId,
    string? ShiftName,
    DateTimeOffset? FirstIn,
    DateTimeOffset? LastOut,
    int WorkedMinutes,
    int LateMinutes,
    int EarlyOutMinutes,
    int OvertimeMinutes,
    AttendanceStatus AttendanceStatus,
    AttendanceSource AttendanceSource,
    bool IsLocked
);

public record CreateRegularisationRequest(
    long EmployeeId,
    DateOnly AttendanceDate,
    DateTimeOffset? RequestedIn,
    DateTimeOffset? RequestedOut,
    AttendanceStatus RequestedStatus,
    string Reason
);

public record RegularisationRequestDto(
    long RegularisationRequestId,
    long EmployeeId,
    DateOnly AttendanceDate,
    DateTimeOffset? RequestedIn,
    DateTimeOffset? RequestedOut,
    AttendanceStatus RequestedStatus,
    string Reason,
    ApprovalStatus ApprovalStatus,
    string? CurrentStepLabel
);

public record CreateOvertimeRequest(
    long EmployeeId,
    DateOnly AttendanceDate,
    int Minutes,
    decimal OvertimeRate
);

public record OvertimeRequestDto(
    long OvertimeRequestId,
    long EmployeeId,
    DateOnly AttendanceDate,
    int Minutes,
    decimal OvertimeRate,
    ApprovalStatus ApprovalStatus,
    string? CurrentStepLabel
);

public record CompOffCreditDto(
    long CompOffCreditId,
    long EmployeeId,
    DateOnly EarnedDate,
    decimal Days,
    DateOnly ExpiryDate,
    decimal AvailedDays,
    decimal AvailableDays
);

public record ApprovalActionRequest(
    ApprovalStepStatus Decision,
    string? Comments
);

public record MonthlyPaidDaysDto(
    long EmployeeId,
    int Year,
    int Month,
    int TotalDays,
    decimal PresentDays,
    decimal HalfDays,
    decimal AbsentDays,
    decimal PaidLeaveDays,
    decimal UnpaidLeaveDays,
    decimal HolidayDays,
    decimal WeeklyOffDays,
    decimal TotalPaidDays
);

File: frontend/libs/time-leave/time-leave-core/src/lib/time-leave.models.ts
export interface LeaveTypeView {
  leaveTypeId: number;
  code: string;
  name: string;
  isPaid: boolean;
  isHalfDayAllowed: boolean;
  isAttachmentRequiredAboveDays?: number | null;
  gender?: string | null;
  isActive: boolean;
}

export interface LeavePolicyView {
  leavePolicyId: number;
  leaveTypeId: number;
  leaveTypeName?: string | null;
  gradeId?: number | null;
  workLocationId?: number | null;
  effectiveFrom: string;
  annualQuota: number;
  accrualKind: string;
  isProratedOnJoining: boolean;
  carryForwardKind: string;
  maxCarryForward?: number | null;
  maxEncashPerYear?: number | null;
  minDaysPerApplication?: number | null;
  maxDaysPerApplication?: number | null;
  noticeDays: number;
  isSandwichRule: boolean;
  canApplyInProbation: boolean;
}

export interface LeaveBalanceView {
  leaveBalanceId: number;
  employeeId: number;
  leaveTypeId: number;
  leaveTypeCode: string;
  leaveTypeName: string;
  leaveYear: number;
  opening: number;
  accrued: number;
  taken: number;
  encashed: number;
  lapsed: number;
  adjusted: number;
  available: number;
}

export interface ApplyLeave {
  employeeId: number;
  leaveTypeId: number;
  fromDate: string;
  toDate: string;
  fromHalf: 'Full' | 'FirstHalf' | 'SecondHalf';
  toHalf: 'Full' | 'FirstHalf' | 'SecondHalf';
  reason: string;
  attachmentKey?: string | null;
}

export interface LeaveApplicationView {
  leaveApplicationId: number;
  employeeId: number;
  leaveTypeId: number;
  leaveTypeCode: string;
  leaveTypeName: string;
  fromDate: string;
  toDate: string;
  fromHalf: string;
  toHalf: string;
  days: number;
  reason: string;
  attachmentKey?: string | null;
  leaveStatus: string;
  approvalStatus: string;
  currentStepLabel?: string | null;
  currentApproverEmployeeId?: number | null;
}

export interface LeaveEncashmentView {
  leaveEncashmentId: number;
  employeeId: number;
  leaveTypeId: number;
  leaveTypeCode: string;
  leaveTypeName: string;
  leaveYear: number;
  days: number;
  encashmentStatus: string;
  approvalStatus: string;
  currentStepLabel?: string | null;
}

export interface ShiftView {
  shiftId: number;
  code: string;
  name: string;
  startTime: string;
  endTime: string;
  breakMinutes: number;
  graceInMinutes: number;
  graceOutMinutes: number;
  halfDayBelowMinutes: number;
  absentBelowMinutes: number;
  isNightShift: boolean;
  isActive: boolean;
}

export interface WeeklyOffPolicyView {
  weeklyOffPolicyId: number;
  name: string;
  mondayRule: string;
  tuesdayRule: string;
  wednesdayRule: string;
  thursdayRule: string;
  fridayRule: string;
  saturdayRule: string;
  sundayRule: string;
  alternateWeeks?: string | null;
  isActive: boolean;
}

export interface DailyAttendanceView {
  dailyAttendanceId: number;
  employeeId: number;
  attendanceDate: string;
  shiftId?: number | null;
  shiftName?: string | null;
  firstIn?: string | null;
  lastOut?: string | null;
  workedMinutes: number;
  lateMinutes: number;
  earlyOutMinutes: number;
  overtimeMinutes: number;
  attendanceStatus: string;
  attendanceSource: string;
  isLocked: boolean;
}

export interface RegularisationRequestView {
  regularisationRequestId: number;
  employeeId: number;
  attendanceDate: string;
  requestedIn?: string | null;
  requestedOut?: string | null;
  requestedStatus: string;
  reason: string;
  approvalStatus: string;
  currentStepLabel?: string | null;
}

export interface PendingApprovalItem {
  approvalStepId: number;
  sequence: number;
  label: string;
  requestKind: string;
  requestId: number;
  employeeId: number;
  employeeName?: string | null;
  reason?: string | null;
  fromDate?: string | null;
  toDate?: string | null;
  days?: number | null;
  attendanceDate?: string | null;
  overtimeMinutes?: number | null;
  createdAt?: string | null;
}

export interface ActApprovalRequest {
  action: 'Approve' | 'Reject' | 'SendBack';
  comments?: string | null;
}

export interface ApplyLeaveSelfRequest {
  leaveTypeId: number;
  fromDate: string;
  toDate: string;
  fromHalf: boolean;
  toHalf: boolean;
  reason?: string | null;
  attachmentKey?: string | null;
}

export interface MobilePunchSelfRequest {
  latitude: number;
  longitude: number;
  deviceInfo?: string | null;
  notes?: string | null;
}

export interface SubmitRegularisationSelfRequest {
  attendanceDate: string;
  requestedIn?: string | null;
  requestedOut?: string | null;
  requestedStatus?: string;
  reason: string;
}

export interface SubmitOvertimeSelfRequest {
  attendanceDate: string;
  minutes: number;
  overtimeRate?: number;
}

